File: ports-py/url-parsing.py
import urllib.parse
import ports
import urllib 

suite = ports.suite("suites/url-parsing-RFC.ports")

@suite.placeholder("url-parse")
def url_parse(env, url_string):
    try:
        return urllib.parse.urlparse(url_string)
    except ValueError as e:
        return e

@suite.placeholder("parse-error?")
def parse_error(env, parse_result):
    return isinstance(parse_result, Exception)

@suite.placeholder("url-scheme")
def url_scheme(env, parse_result: urllib.parse.ParseResult):
    return parse_result.scheme or ""

@suite.placeholder("url-authority")
def url_authority(env, parse_result: urllib.parse.ParseResult):
    return parse_result.netloc or ""

#
# Running
#

suite.run()

File: ports-js/url-parsing.js
import { suite } from './ports.js';
import { URL } from 'url';

const urlParsingSuite = suite("suites/url-parsing-RFC.ports");

urlParsingSuite.placeholder("url-parse", (env, urlString) => {
    try {
        return new URL(urlString);
    } catch (e) {
        return e;
    }
});

urlParsingSuite.placeholder("parse-error?", (env, parseResult) => {
    return parseResult instanceof Error;
});

urlParsingSuite.placeholder("url-scheme", (env, parseResult) => {
    return parseResult instanceof URL ? parseResult.protocol.replace(':', '') : "";
});

urlParsingSuite.placeholder("url-authority", (env, parseResult) => {
    return parseResult instanceof URL ? parseResult.host : "";
});

// Running
urlParsingSuite.run();
